libc/math:Update testcases for math library

drashti304 opened the pull request:
Update the testcases to cover all the corner cases
and update the respective function implementation wherever required.

Signed-off-by: Drashti Parikh <p.drashti@partner.samsung.com>

seinfra commented:
Target : [38b08dfa9488ff4deba1d1a62280f22223ad93f5] - Code Rule Check Result:
apps/examples/testcase/tc_common.h:45: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:76: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:88: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:100: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:112: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:124: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:136: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/tc_common.h:151: ERROR: [SPC_M_SEP] space required after that close brace '}'
apps/examples/testcase/tc_common.h:154: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'

seinfra commented:
Target : [38b08dfa9488ff4deba1d1a62280f22223ad93f5] - Code Rule Check (C++) OK.

jeongchanKim commented:
Please split the commit about updating math library and modifying tcs.

seinfra commented:
Target : [ec4541e0127a367eceb004521bc99de4081cf895] - Code Rule Check Result:
apps/examples/testcase/tc_common.h:45: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:76: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:88: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:100: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:112: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:124: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:136: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
apps/examples/testcase/tc_common.h:154: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)

seinfra commented:
Target : [ec4541e0127a367eceb004521bc99de4081cf895] - Code Rule Check (C++) OK.

drashti304 commented:
> Please split the commit about updating math library and modifying tcs.

Separated two commits for math library functions and testcases

jeongchanKim commented:
> > Target : [[ec4541e](https://github.com/Samsung/TizenRT/commit/ec4541e0127a367eceb004521bc99de4081cf895)] - Code Rule Check Result:
> > apps/examples/testcase/tc_common.h:45: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:76: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:88: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:112: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:124: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:136: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:154: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> 
> Last Code Rule Check Result and this result is conflicting.
> How to solve this?



> > Target : [[ec4541e](https://github.com/Samsung/TizenRT/commit/ec4541e0127a367eceb004521bc99de4081cf895)] - Code Rule Check Result:
> > apps/examples/testcase/tc_common.h:45: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:76: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:88: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:112: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:124: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:136: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> > apps/examples/testcase/tc_common.h:154: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxB)
> 
> Last Code Rule Check Result and this result is conflicting.
> How to solve this?

Can you change the space to (void)0 like below?
`#define TC_ASSERT(api_name, var) TC_ASSERT_CLEANUP(api_name, var, ((void)0))`

yashwanth686007 commented on apps/examples/testcase/tc_common.h:
```diff
@@ -133,7 +133,25 @@ extern int total_fail;
-#define TC_ASSERT_LEQ(api_name, var, ref) TC_ASSERT_LEQ_CLEANUP(api_name, var, ref, )
+#define TC_ASSERT_LEQ(api_name, var, ref) TC_ASSERT_LEQ_CLEANUP(api_name, var, ref,)
+
+#define TC_ASSERT_NAN_CLEANUP(api_name, var, ref, freeResource) \
```
Is it not possible to compare using boolean check values instead of this ?

drashti304 replied:
If we compare using boolean  check then also we need to print the result, so instead of duplicating the same effort in all the functions, I have implemented the common API for that.
And I think  we follow the same method for other checks also like LEQ and EQ for the other values.

drashti304 replied:
Updated the logic to eliminate the TC_ASSERT_NAN macro

yashwanth686007 commented on lib/libc/math/lib_asin.c:
```diff
@@ -64,6 +64,14 @@ double asin(double x)
+		return NAN;
+	}
+
+	if (x < -1.0 || x > 1.0) {
```
Try to reduce the redundant if loops by clubbing the conditions.

drashti304 replied:
Modified

yashwanth686007 commented on lib/libc/math/lib_asinf.c:
```diff
@@ -61,6 +61,14 @@ float asinf(float x)
 
 	y = 0;
 
+	if (isnan(x)) {
```
Same as above.

drashti304 replied:
Modified

yashwanth686007 commented on os/include/tinyara/math.h:
```diff
@@ -116,6 +116,7 @@
 #define INFINITY    (1.0/0.0)
 #define NAN         (0.0/0.0)
 #define HUGE_VAL    INFINITY
+#define PI	    3.141592653f
```
I think PI is already is defined in fixedmath.h header.

yashwanth686007 commented on lib/libc/math/lib_logl.c:
```diff
@@ -62,30 +70,56 @@ long double logl(long double x)
 	long double y_old;
 	long double ey;
 	long double epsilon;
+	int iter, relax_factor;
```
Please declare variable in separate lines as per coding guidelines.

drashti304 replied:
Modified

zhouxinhe commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -49,16 +49,22 @@
 		ret_val[acos_idx] = acos(in_val[acos_idx]);
-		TC_ASSERT_LEQ("acos", fabs(sol_val[acos_idx] - ret_val[acos_idx]), FLT_EPSILON);
+		if (!(isnan(sol_val[acos_idx]) && isnan(ret_val[acos_idx]))) {
+			if (isnan(ret_val[acos_idx]) || isnan(sol_val[acos_idx])) {
```
gggg

drashti304 commented:
> @drashti304 Could you update this with @yashwanth686007 's comment?
> Additionally, please write what it is in commit title.
> I think you can write your description at title and then, write what corner case is in description.

I have addressed the comment of yashwanth, and modified the commit message as requested.

seinfra commented:
Target : [db7530716b3479ac65bb775ec21e9b3e19fd3595] - Code Rule Check (C++) OK.

seinfra commented:
Target : [db7530716b3479ac65bb775ec21e9b3e19fd3595] - Code Rule Check Result:
apps/examples/testcase/le_tc/kernel/tc_libc_math.c:339: ERROR: [SPC_M_OPR] spaces required around that '=' (ctx:VxV)

yashwanth686007 commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -257,16 +322,24 @@ static void tc_libc_math_asinf(void)
+			compute_val = DUMMY_VALUE;
+		}
+
+		TC_ASSERT_LEQ("asin", compute_val, ref_val);
```
"asin" -> "asinh"

yashwanth686007 commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -1125,12 +1444,18 @@ static void tc_libc_math_expl(void)
 	long double ret_val[SIZE(sol_val, long double)];
 	int expl_idx;
+	long double compute_val;
+	long double ref_val = FLT_EPSILON;
```
FLT_EPSILON -> DBL_EPSILON

yashwanth686007 commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -1151,12 +1476,18 @@ static void tc_libc_math_exp2(void)
 	double ret_val[SIZE(sol_val, double)];
 	int exp2_idx;
+	double compute_val;
+	double ref_val = FLT_EPSILON;
```
Same as above

yashwanth686007 commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -2046,6 +2549,7 @@ static void tc_libc_math_log(void)
 
 	for (log_idx = 0; log_idx < SIZE(sol_val, double); log_idx++) {
 		ret_val[log_idx] = log(in_val[log_idx]);
+		//printf("ret_val=%lf for in_val=%lf\n", ret_val[log_idx], in_val[log_idx]);
```
Please delete the unnecessary lines.

seinfra commented:
Target : [4484d2b03edfcf14b78c1c41f6103f9a8c549499] - Code Rule Check OK.

seinfra commented:
Target : [4484d2b03edfcf14b78c1c41f6103f9a8c549499] - Code Rule Check (C++) OK.

seinfra commented:
Target : [8ac0dd0c273c93dec5640b46e67bcc6d97ccf96c] - Code Rule Check OK.

seinfra commented:
Target : [8ac0dd0c273c93dec5640b46e67bcc6d97ccf96c] - Code Rule Check (C++) OK.

seinfra commented:
Target : [5f7930bca4fa1f98279ed16ce8de45c9c56549f7] - Code Rule Check (C++) OK.

sunghan-chang commented:
@yashwanth686007 Could you check this with your comment and then approve?
@drashti304 @yashwanth686007 You guys are better than me for this. After getting approval from @yashwanth686007 or @jeongchanKim or someone else, I will merge this. (Sorry but no time to see)

thapav commented on lib/libc/math/lib_floorl.c:
```diff
@@ -57,11 +57,16 @@
 		x -= 1.0;
 	}
 
+	modfl(x, &x);
```
Also, should void typecast be added to the modf, modff, modfl calls in all the floor API implementations?
As the modf() will return the integer part of X
What is your opinion ?

NilabhShant commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -179,16 +221,24 @@ static void tc_libc_math_acoshl(void)
 	for (acosl_idx = 0; acosl_idx < SIZE(sol_val, long double); acosl_idx++) {
 		ret_val[acosl_idx] = acosl(in_val[acosl_idx]);
-		TC_ASSERT_LEQ("acosl", fabsl(sol_val[acosl_idx] - ret_val[acosl_idx]), FLT_EPSILON);
+		compute_val = fabs(sol_val[acosl_idx] - ret_val[acosl_idx]);
```
No, just a typing mistake. Rectified now

NilabhShant commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -75,16 +85,24 @@ static void tc_libc_math_acos(void)
 	float ret_val[SIZE(in_val, float)];
 	int acosf_idx;
+	float compute_val;
+	float ref_val = FLT_EPSILON;
```
Modified

NilabhShant commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -3352,12 +4233,18 @@ static void tc_libc_math_y1f(void)
+			compute_val = DUMMY_VALUE;
 		}
+
+		TC_ASSERT_LEQ("y1f_idx", compute_val, ref_val);
```
Rectified

seinfra commented:
Target : [9f974b5e9e22bf1809ee0264f0b7a4eb9aa53ac0] - Code Rule Check (C++) OK.

seinfra commented:
Target : [9f974b5e9e22bf1809ee0264f0b7a4eb9aa53ac0] - Code Rule Check OK.

NilabhShant commented on lib/libc/math/lib_round.c:
```diff
@@ -40,7 +40,14 @@
 double round(double x)
 {
-	double f = modf(x, &x);
+	double f ;
```
Ok sure, Will be fix in next patch

seinfra commented:
Target : [99f980a878660d58c41166b81312c5cd48dca90f] -

seinfra commented:
Target : [99f980a878660d58c41166b81312c5cd48dca90f] -

seinfra commented:
Target : [65bbe6bce81e9eb56932b31f104a7ac14d60848e] - Code Rule Check OK.

seinfra commented:
Target : [65bbe6bce81e9eb56932b31f104a7ac14d60848e] - Code Rule Check (C++) OK.

sunghan-chang commented:
@NilabhShant Could you let me know why you plan to fix all of comment with next patch, not here?

NilabhShant commented:
@sunghan-chang Next patch means next push to the same PR not a different commit.

sunghan-chang commented:
@NilabhShant Could you let me know the schedule when you fix?

NilabhShant commented:
@sunghan-chang its already done .

sunghan-chang commented:
@kishore-sn @thapav Could you review this for what you left comments?

sunghan-chang commented:
@NilabhShant After approval from reviewer who left comments, I will merge this.

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -1650,19 +2046,35 @@ static void tc_libc_math_frexp(void)
+			compute_val = ONE;
+		}
+
+		TC_ASSERT_EQ("frexp", compute_val, ref_val);
```
frexp -> frexpf

drashti304 replied:
Modified

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -2495,17 +3073,22 @@ static void tc_libc_math_remquo(void)
+			compute_val = ONE;
 		}
+
+		TC_ASSERT_LEQ("acos", compute_val, FLT_EPSILON);
```
acos >> remquo

drashti304 replied:
Modified

kishore-sn commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -179,16 +218,23 @@ static void tc_libc_math_acoshl(void)
-	const long double in_val[] = { 0.64, -0.4, ZERO, -1, 1 };
-	const long double sol_val[] = { 0.87629806116834, 1.9823131728624, 1.5707963267949, 3.1415926535898, 0 };
+	const double in_val[] = { 0.64, -0.4, ZERO, -1, 1, 1.1, -2, INFINITY, -INFINITY, NAN };
+	const double sol_val[] = { 0.876298061168341, 1.982313172862385, 1.570796326794897, 3.141592653589793, 0, NAN, NAN, NAN, NAN, NAN };
```
Please change to long double.

drashti304 replied:
Changed

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -3088,16 +3790,23 @@ static void tc_libc_math_tan(void)
+	const double sol_val[] = { 1, 0.54630248984379, -0.54630248984379, -3.3805160522461, 0.6483603715897, 3.3805131912231, -0.6483607292175, ZERO, NAN, NAN, NAN };
 	double ret_val[SIZE(in_val, double)];
 	int tanf_idx;
+	double compute_val;
```
Kindly change all double to float

drashti304 replied:
Changed

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -3140,16 +3856,23 @@ static void tc_libc_math_tanh(void)
+	const double sol_val[] = { 0.65579420263267, 0.46211715726001, -0.46211715726001, 0.9999092042626, 0.99999999587769, -0.9999092042626, -0.99999999587769, ZERO, 1.0, -1.0, NAN };
 	double ret_val[SIZE(in_val, double)];
 	int tanhf_idx;
+	double compute_val;
```
Kindly change all double to float

drashti304 replied:
Modified the code accordingly

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -3166,16 +3889,23 @@ static void tc_libc_math_tanhf(void)
+	const double sol_val[] = { 0.65579420263267, 0.46211715726001, -0.46211715726001, 0.9999092042626, 0.99999999587769, -0.9999092042626, -0.99999999587769, ZERO, 1.0, -1.0, NAN };
 	double ret_val[SIZE(in_val, double)];
 	int tanhl_idx;
+	double compute_val;
```
Kindly change all
double > long double

drashti304 replied:
Modified

thapav commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -3192,16 +3922,23 @@ static void tc_libc_math_tanhl(void)
+	const double sol_val[] = { 1, 0.54630248984379, -0.54630248984379, -3.3805150062466, 0.64836082745909, 3.3805150062466, -0.64836082745909, ZERO, NAN, NAN, NAN };
 	double ret_val[SIZE(in_val, double)];
 	int tanl_idx;
+	double compute_val;
```
Kindly change all
double > long double

drashti304 replied:
Modified

drashti304 commented on apps/examples/testcase/le_tc/kernel/tc_libc_math.c:
```diff
@@ -2327,17 +2880,22 @@ static void tc_libc_math_nexttoward(void)
 {
-	const float in_val1[] = { VAL1, VAL1, ZERO, -VAL1, -VAL2 , VAL1, NAN };
+	const float in_val1[] = { VAL1, VAL1, ZERO, -VAL1, -VAL2, VAL1, NAN };
 	const long double in_val2[] = { VAL1, VAL2, VAL1, -VAL2, ZERO, NAN, VAL2 };
```
According to the definition of nexttoeardf, in_val2 is long double only.

seinfra commented:
Target : [fe2e99e8d94264d731fbc7a145e5c21273109e17] - Code Rule Check OK.

seinfra commented:
Target : [fe2e99e8d94264d731fbc7a145e5c21273109e17] - Code Rule Check (C++) OK.

seinfra commented:
Target : [cd44ea1eaeaaed2d6c05d871039bb4ab129658db] - Code Rule Check (C++) OK.

seinfra commented:
Target : [cd44ea1eaeaaed2d6c05d871039bb4ab129658db] - Code Rule Check OK.

drashti304 commented:
@sunghan-chang Kindly merge the patch once you review the changes.

sunghan-chang commented:
@drashti304 @NilabhShant I just see commit titles. 
1. libc/math:Update testcases for math library to cover corner cases
    Words before `:` mean a location of modification. Second commit has changes on testcase, not libc.math.
    Let's use `testcase/math: cover conrner cases for math library`.
2. build/configs/qemu/tc_64k: Modify defconfig
   It is difficult to know what you do from title. 
   1. Let's use `build/configs/qemu/tc_64k/defconfig: disable math tc defaultly`.
   2. That is a tc defconfig. Main purpose is executing tc. How about disable another codes like tash command?

sunghan-chang commented:
@drashti304 @NilabhShant Because you already got approvals from another reviewers, I have no plan to review source codes. I just want to modify titles.

seinfra commented:
Target : [3eeb4278648691f24d193ced80296a62f8b6c557] - Code Rule Check (C++) OK.

seinfra commented:
Target : [3eeb4278648691f24d193ced80296a62f8b6c557] - Code Rule Check OK.

seinfra commented:
Target : [92d1a774b380309a1bd57d010f2426e540687725] -

seinfra commented:
Target : [1bb861cee8995102be50d0616f361b017c349f25] - Code Rule Check OK.

seinfra commented:
Target : [1bb861cee8995102be50d0616f361b017c349f25] - Code Rule Check (C++) OK.

seinfra commented:
Target : [e873bc7b8056a64f50dabbe69ab7a385d4baea71] - Code Rule Check OK.

seinfra commented:
Target : [e873bc7b8056a64f50dabbe69ab7a385d4baea71] - Code Rule Check (C++) OK.

seinfra commented:
Target : [d5df75020f78bad401217c3f64862b47997f23a8] - Code Rule Check OK.

seinfra commented:
Target : [d5df75020f78bad401217c3f64862b47997f23a8] - Code Rule Check (C++) OK.

seinfra commented:
Target : [679fba99c8d25cb02024806e6460782c3d9efc6a] - Code Rule Check (C++) OK.

seinfra commented:
Target : [679fba99c8d25cb02024806e6460782c3d9efc6a] - Code Rule Check OK.

drashti304 commented:
> @drashti304 @NilabhShant I just see commit titles.
> 
>     1. libc/math:Update testcases for math library to cover corner cases
>        Words before `:` mean a location of modification. Second commit has changes on testcase, not libc.math.
>        Let's use `testcase/math: cover conrner cases for math library`.
> 
>     2. build/configs/qemu/tc_64k: Modify defconfig
>        It is difficult to know what you do from title.
>        
>        1. Let's use `build/configs/qemu/tc_64k/defconfig: disable math tc defaultly`.
>        2. That is a tc defconfig. Main purpose is executing tc. How about disable another codes like tash command?

I have modified the commit message to testcase/math as well as disable tash commands instead of math tc as suggested.

Please merge the PR once you review the changes.
